plan/eliminate_projection.go

```diff
@@ -14,50 +14,229 @@
 	}
 	for i, expr := range p.Exprs {
 		col, ok := expr.(*expression.Column)
-		if !ok {
-			return false
-		}
-		if col.FromID != child.Schema().Columns[i].FromID || col.Position != child.Schema().Columns[i].Position {
+		if !ok || !col.Equal(child.Schema().Columns[i], nil) {
 			return false
 		}
 	}
 	return true
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string(origin.HashCode())]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
+	switch expr := origin.(type) {
+	case *expression.Column:
+		resolveColumnAndReplace(expr, replace)
+	case *expression.ScalarFunction:
+		for _, arg := range expr.GetArgs() {
+			resolveExprAndReplace(arg, replace)
+		}
+	}
+}
+
+// eliminatePhysicalProjection should be called after physical optimization to eliminate the redundant projection
+// left after logical projection elimination.
+func eliminatePhysicalProjection(p PhysicalPlan, replace map[string]*expression.Column) PhysicalPlan {
+	children := make([]Plan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		children = append(children, eliminatePhysicalProjection(child.(PhysicalPlan), replace))
+	}
+	p.SetChildren(children...)
+
+	for _, dst := range p.Schema().Columns {
+		resolveColumnAndReplace(dst, replace)
+	}
+	for _, key := range p.Schema().Keys {
+		for _, keyCol := range key {
+			resolveColumnAndReplace(keyCol, replace)
+		}
+	}
+	p.replaceExprColumns(replace)
+
+	proj, isProj := p.(*Projection)
+	if !isProj || !canProjectionBeEliminatedStrict(proj) {
+		return p
+	}
+	child := p.Children()[0]
+	exprs := proj.Exprs
+	childCols := child.Schema().Columns
+	for i, parentColumn := range proj.Schema().Columns {
+		col, _ := exprs[i].(*expression.Column)
+		col.DBName = parentColumn.DBName
+		col.TblName = parentColumn.TblName
+		col.ColName = parentColumn.ColName
+		childCols[i].ColName = parentColumn.ColName
+		replace[string(parentColumn.HashCode())] = col
+	}
+	RemovePlan(p)
+	return child.(PhysicalPlan)
+}
+
+type projectionEliminater struct {
+}
+
+// optimize implements the logicalOptRule interface
+func (pe *projectionEliminater) optimize(lp LogicalPlan, _ context.Context, _ *idAllocator) (LogicalPlan, error) {
+	root := pe.eliminate(lp, make(map[string]*expression.Column), false)
+	return root.(LogicalPlan), nil
+}
+
+// eliminate eliminates the redundant projection in a logical plan.
+// The order of output plan's schema columns can be changed if "haveProjection" is true.
+func (pe *projectionEliminater) eliminate(p LogicalPlan, replace map[string]*expression.Column, haveProjection bool) LogicalPlan {
+	proj, isProj := p.(*Projection)
+	children := make([]Plan, 0, len(p.Children()))
+	_, isUnion := p.(*Union)
+	for _, child := range p.Children() {
+		children = append(children, pe.eliminate(child.(LogicalPlan), replace, !isUnion && (haveProjection || isProj)))
```

Comment:
Well, I think there are two points that can be improved:
1. haveProjection is not a good name.
2. !isUnion && (haveProjection || isProj) is not readable. 
The following style is more acceptable:
```
var childFlag = haveProjection
if isProj {
    childFlag = true
} else if _, isUnion := p.(*Union); isUnion {
    childFlag = false
}
```